Parallelizing rinex2snr and gnssir

aaryan-rampal opened the pull request:
I use this utility to analyze soil data from multiple sites for a few decades. I have noticed that there doesn't seem to be any dependency of year n by year n+1, so I was wondering why not create parallel threads which run these two utilities side by side for multiple years. 

For now, it's just a small change to rinex2snr.py with the changes. If you rinex2snr in the terminal, it should create a maximum of 10 concurrently running threads. If this idea is approved, I can work on adding docstrings, comments, and terminal options to turn on and off parallelisation alongside an arg for how many threads at once should be run.

kristinemlarson commented:
Thanks for working on this. I'll take a look.

kristinemlarson commented:
i think (?) there are a couple issues.  one relates to downloading and accessing data from GNSS archives. Doing that for 10 files at the same time might not work for everyone's internet connections.  But I don't see why it could not be an option.

aaryan-rampal commented:
I added the options -par into gnssir and rinex2snr. The default is no multiprocessing occurs, and it calls the proces_year function for each year in a for loop.

I was having problems with rinex2snr_cl calling run_rinex2snr. I think it imported the version of gnssrefl.rinex2snr I had downloaded using pip, which did not have the par argument. I had to force it to import the local file version rinex2snr which I had updated. I don't know of a better fix to this.

Also, what is year_st in gnssir.

kristinemlarson commented:
i have not looked at all these threads - and i won't be working today.  i
am pretty sure year_st is the starting year of your analysis request.
originally i made things so that you could only do a single year, but you
could request a starting and ending doy.  a colleague did not like that -
and added some edits that allowed multiple years. not 100% sure, but but
that is my memory.  i will try to look thru your changes tomorrow.  the
code has evolved as my understanding of python has improved - but there are
still places that make me cringe when i look back at them.  my general
feeling is to not change working code unless there is a good reason.
sometimes the reason is simply that it is easier to maintain if the change
is made. and i'm good with that one. another reason is to make the code
faster, as you are trying to do.  so i am not against change - but it has
to be sustainable.  and the change should not make it harder to maintain.
another issue is that i am having the code tested externally in hte next
couple weeks and i am also teaching a class that uses gnssrefl on march
6/7.  so i don't want to break the code in the next two weeks.  but i think
we can work around those constraints.


On Fri, Feb 23, 2024 at 10:10 PM Aaryan Rampal ***@***.***>
wrote:

> I added the options -par into gnssir and rinex2snr. The default is no
> multiprocessing occurs, and it calls the proces_year function for each year
> in a for loop.
>
> I was having problems with rinex2snr_cl calling run_rinex2snr. I think it
> imported the version of gnssrefl.rinex2snr I had downloaded using pip,
> which did not have the par argument. I had to force it to import the local
> file version rinex2snr which I had updated. I don't know of a better fix to
> this.
>
> Also, what is year_st in gnssir.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kristinemlarson/gnssrefl/pull/245#issuecomment-1961993816>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFSDAPE6JPZMSMUQB72I35TYVEAWFAVCNFSM6AAAAABDV7WWVGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNRRHE4TGOBRGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Kristine M. Larson
Professor Emerita, University of Bonn
***@***.***
https://www.kristinelarson.net <http://www.kristinelarson.net>
https://gnss-reflections.org
https://github.com/kristinemlarson

Are you waiting for your code to be perfect?

kristinemlarson commented:
I have some time today - so I will start going through this.  Thank you again for posting it!

kristinemlarson commented:
i can see that happening.  maybe because i am suboptimally doing those
downloads.
like maybe they should be in a try?

i have not made much progress on your code because of preparing for short
course
next week teaching gnssrefl.

On Fri, Mar 1, 2024 at 8:17 AM Aaryan Rampal ***@***.***>
wrote:

> I've noticed a bug. Sometimes, gnssir wants you to enter input (such as
> whether or not to overwrite an existing file). When that happens, that
> process stops, waiting for an input from the user but gets lost in the logs
> of the other processes running. I will see if I can't find a way to pause
> all the processes when an input is required from the user, get the input,
> and then resume all the processes.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kristinemlarson/gnssrefl/pull/245#issuecomment-1972637736>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFSDAPHP6GTP7RC7R2M4ARDYWATPHAVCNFSM6AAAAABDV7WWVGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZSGYZTONZTGY>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Kristine M. Larson
Professor Emerita, University of Bonn
***@***.***
https://www.kristinelarson.net <http://www.kristinelarson.net>
https://gnss-reflections.org
https://github.com/kristinemlarson

Are you waiting for your code to be perfect?

kristinemlarson commented:
or maybe just add a force flag (-f) when gunzipping or similar.

On Fri, Mar 1, 2024 at 9:59 AM Kristine Larson ***@***.***>
wrote:

> i can see that happening.  maybe because i am suboptimally doing those
> downloads.
> like maybe they should be in a try?
>
> i have not made much progress on your code because of preparing for short
> course
> next week teaching gnssrefl.
>
> On Fri, Mar 1, 2024 at 8:17 AM Aaryan Rampal ***@***.***>
> wrote:
>
>> I've noticed a bug. Sometimes, gnssir wants you to enter input (such as
>> whether or not to overwrite an existing file). When that happens, that
>> process stops, waiting for an input from the user but gets lost in the logs
>> of the other processes running. I will see if I can't find a way to pause
>> all the processes when an input is required from the user, get the input,
>> and then resume all the processes.
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kristinemlarson/gnssrefl/pull/245#issuecomment-1972637736>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AFSDAPHP6GTP7RC7R2M4ARDYWATPHAVCNFSM6AAAAABDV7WWVGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZSGYZTONZTGY>
>> .
>> You are receiving this because you commented.Message ID:
>> ***@***.***>
>>
>
>
> --
> Kristine M. Larson
> Professor Emerita, University of Bonn
> ***@***.***
> https://www.kristinelarson.net <http://www.kristinelarson.net>
> https://gnss-reflections.org
> https://github.com/kristinemlarson
>
> Are you waiting for your code to be perfect?
>


-- 
Kristine M. Larson
Professor Emerita, University of Bonn
***@***.***
https://www.kristinelarson.net <http://www.kristinelarson.net>
https://gnss-reflections.org
https://github.com/kristinemlarson

Are you waiting for your code to be perfect?

kristinemlarson commented:
my short course is over - so i can give this some time now.

kristinemlarson commented:
@timdittmann and @aaryan-rampal  I would like to start trying this PR out on my local machine. But I am not 100% sure how to install this PR locally.  I was thinking there would be a branch - but there doesn't seem to be?  Anyway, I think the changes look reasonable but would like to run it locally before accepting it.  If either of you know, please send me the instructions.  I think the gnssir changes are cleaner (i.e. they won't have issues with downloads from external archives) but the PR are together, so I am fine with trying to test them both at the same time.

kristinemlarson commented:
here 
https://github.com/aaryan-rampal/gnssrefl.git

it does have a place where I could create a new fork. Maybe that is what I should do? ?? Then???

kristinemlarson commented:
once we get this working it would be nice to have it work on sub-years!

aaryan-rampal commented:
Sorry I have been a bit busy and not been able to work on this much. This is my first PR so I did not know how to handle the branches. I will make different branches for the gnssir and rinex2snr changes, is that what you mean?

For the problem with overwriting, I am working on writing code which sees if any process receives stderror output and if it does, it can pause all the other threads until action has been received. If that doesn't work, the -f flag is a backup option. 

How do you envision it working for sub years? Batches of days at a time which scale with how long the user wants? It might be possible to see how many days the user wants (seperate function), divide it by parallel processes wanted and then give each process the same amount of work. I wonder if it might have bugs regarding making directories if two processes for the same years but different dates are working together. 

I will be working on these features tomorrow

aaryan-rampal commented:
There are two different branches of this code which isolate the two different files changed. https://github.com/aaryan-rampal/gnssrefl/tree/gnssir and https://github.com/aaryan-rampal/gnssrefl/tree/rinex2snr

kristinemlarson commented:
*How do you envision it working for sub years? Batches of days at a time
which scale with how long the user wants? It might be possible to see how
many days the user wants (seperate function), divide it by parallel
processes wanted and then give each process the same amount of work. *

this was what i was thinking.  like if they asked for 100 days in the year
2023, split it into ten jobs with
ten days.  that would be sweet.  sure, going over the year boundary is a
slight issue - but only
because i wrote the code in such a silly way.  for now i've been working on
a branch that keeps your new
code the same (if it ain't broke, don't fix it) but has a separate function
where the data
handling would be a little smarter (on my end, not your end).  your new
code clearly would be great
for people that wanted to analyze five years of data at the same time.  and
for sure there are people
that want to do that - but sometimes people only want a year or so, and i
think it would nice
to get a subyear option also working.

i had no time last week because of the class, but i was able to work on
this a little bit yesterday.  it is nice
to do something new for a change.  also made a demo.gif for my README file
;-)

k.




On Wed, Mar 13, 2024 at 8:12 AM Aaryan Rampal ***@***.***>
wrote:

> Sorry I have been a bit busy and not been able to work on this much. This
> is my first PR so I did not know how to handle the branches. I will make
> different branches for the gnssir and rinex2snr changes, is that what you
> mean?
>
> For the problem with overwriting, I am working on writing code which sees
> if any process receives stderror output and if it does, it can pause all
> the other threads until action has been received. If that doesn't work, the
> -f flag is a backup option.
>
> How do you envision it working for sub years? Batches of days at a time
> which scale with how long the user wants? It might be possible to see how
> many days the user wants (seperate function), divide it by parallel
> processes wanted and then give each process the same amount of work. I
> wonder if it might have bugs regarding making directories if two processes
> for the same years but different dates are working together.
>
> I will be working on these features tomorrow
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kristinemlarson/gnssrefl/pull/245#issuecomment-1993704394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFSDAPEAGLB3LRDCRZTMXX3YX735TAVCNFSM6AAAAABDV7WWVGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJTG4YDIMZZGQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Kristine M. Larson
Professor Emerita, University of Bonn
***@***.***
https://www.kristinelarson.net <http://www.kristinelarson.net>
https://gnss-reflections.org
https://github.com/kristinemlarson

Are you waiting for your code to be perfect?

kristinemlarson commented:
Just got it to work for simultaneous processes for < one year!  it would have taken much less time if i had read the instructions for the partial function first.

kristinemlarson commented:
ok - now i have it working across years or within years.  i switched the way I defined the lists to use Modified Julian dates to keep my sanity - and then converted back to year/doy at the end.  yippee!  

i have not pushed it to github yet - but when i do it is named gnssir2.  

tomorrow I will start looking at the rinex2snr part.  

k.

kristinemlarson commented:
just did a test of our current rinex2snr, downloading highrate data from one of the slow archives (but not the slowest) in two windows. and it does not really help make things much faster (143 sec instead of 156 sec).  for a fast archive, and smaller files, and testing a whole year, perhaps you can start to see something.  for analyzing your own data, so no download issues, i am sure doing in parallel would be a big savings.  i will keep thinking about it.

aaryan-rampal commented:
This branch is not actually updated anymore, I have created 2 new PRs which contain the relevant gnssir and rinex2snr branches

kristinemlarson commented:
thanks much.  the missing function for modified julian day (maybe you
already say) was in a new version of gps.py

k.

On Wed, Mar 13, 2024 at 6:54 PM Aaryan Rampal ***@***.***>
wrote:

> This branch is not actually updated anymore, I have created 2 new PRs
> which contain the relevant gnssir and rinex2snr branches
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kristinemlarson/gnssrefl/pull/245#issuecomment-1995179045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFSDAPAPMII6Z2IDFD7O3QDYYCHE5AVCNFSM6AAAAABDV7WWVGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJVGE3TSMBUGU>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Kristine M. Larson
Professor Emerita, University of Bonn
***@***.***
https://www.kristinelarson.net <http://www.kristinelarson.net>
https://gnss-reflections.org
https://github.com/kristinemlarson

Are you waiting for your code to be perfect?
